### PR TITLE
SCTX-2115 CA-182595: Force lifecycle operations must cancel clean ones

### PR DESCRIPTION
**Pull request by thomassa**
This cream-bugfix pull-request cherry-picks the pull-request #44 commits
from clearwater-sp1-bugfix.

These came originally from xapi-project/xen-api#2272 and
xapi-project/xen-api#2400.

### CODE REVIEW

**Comment by simonjbeaumont** [comment]
@xen-git retest this please.

**Comment by simonjbeaumont** [comment]
Looks like a proper failure:

```
15:09:08 + ocamlfind ocamlopt -package oclock,xml-light2,cdrom,pciutil,sexpr,log,stunnel,http-svr,rss,xen-utils,netdev,tapctl,vhd,rpc-light,nbd,xenstore,xenctrl,xenctrl,xenctrlext,xenstore,xenstoreext,cpuid -g -dtypes -thread -warn-error +a-4-6-9-27-28-29 -ccopt -fPIC -I ../idl -I ../idl/ocaml_backend -I ../autogen -I ../database/ -I ../xenops -I ../xva -I ../util -I ../auth -I ../license -I ../client_records -I ../rfb -I ../gpg -I ../network -I ../rrdd/interface -I .. -c xapi_vm.ml
15:09:08 File "xapi_vm.ml", line 722, characters 3-32:
15:09:08 Warning 5: this function application is partial,
15:09:08 maybe some arguments are missing.
15:09:08 File "xapi_vm.ml", line 725, characters 3-33:
15:09:08 Warning 5: this function application is partial,
15:09:08 maybe some arguments are missing.
15:09:08 File "xapi_vm.ml", line 1, characters 0-1:
15:09:08 Error: The implementation xapi_vm.ml
15:09:08        does not match the interface xapi_vm.cmi:
15:09:08        Values do not match:
15:09:08          val request_rdp_on :
15:09:08            __context:Context.t -> vm:[ `VM ] Ref.t -> vm:'a Ref.t -> unit
15:09:08        is not included in
15:09:08          val request_rdp_on : __context:Context.t -> vm:API.ref_VM -> unit
15:09:09 *** omake: 2927/2946 targets are up to date
15:09:09 *** omake: failed (3 min 03.22 sec, 504/574 scans, 621/1566 rules, 1879/4952 digests)
15:09:09 *** omake: targets were not rebuilt because of errors:
15:09:09    ocaml/xapi/xapi_vm.cmx
15:09:09       depends on: ocaml/xapi/xapi_vm.ml
15:09:09    ocaml/xapi/xapi_vm.o
15:09:09       depends on: ocaml/xapi/xapi_vm.ml
15:09:09 make[4]: *** [all] Error 2
```

**Comment by thomassa** [comment]
Yes, the function request_rdp_on is expecting
`__context:Context.t -> vm:API.ref_VM` but is being passed
`__context:Context.t -> vm:[ `VM ] Ref.t -> vm:'a Ref.t` somehow.

File: ocaml/xapi/xapi_xenops.ml [hunk 1]
```diff
@@ -1835,25 +1835,25 @@ let sync __context x =
...
 			let id = id_of_vm ~__context ~self in
 			debug "xenops: VM.unpause %s" id;
 			let dbg = Context.string_of_task __context in
 			Client.VM.unpause dbg id |> sync_with_task __context;
 			Events_from_xenopsd.wait dbg id ();
-			assert (Db.VM.get_power_state ~__context ~self = `Running)
+			Xapi_vm_lifecycle.assert_power_state_is ~__context ~self ~expected:`Running
 		)
 
 let request_rdp ~__context ~self enabled =
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_xenops.ml:1859]: You need to add a `~vm:self` to the `transform_xenops_exn` line below...

**Comment by simonjbeaumont** [comment]
You need to make the following change to line 1860 of ocaml/xapi/xapi_xenops.ml:

``` diff
-   transform_xenops_exn ~__context
+   transform_xenops_exn ~__context ~vm:self
```

I guess it got missed in the cherry pick because these functions weren't there in the original patch?

**Comment by thomassa** [comment]
Thank you: well spotted.
That's right: `request_rdp_on` and `request_rdp_off` are in cream but not in clearwater, so applying the change from clearwater to cream doesn't change them. I'll fix that up.

**Comment by thomassa** [comment]
I've made that change, and squashed it into the "Fix laundering of Bad_power_state exn" commit.

**Comment by robhoes** [comment]
This now passed the tests. Merging...

**Pull request merged**